pkg/util/sliding_window.go

```diff
@@ -0,0 +1,218 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package util
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// PollingFunc is polled at specified intervals to collect the value
+// that should be included in the slidingWindow data
+type PollingFunc func() (value float64)
+
+// CallbackFunc is invoked when the polling function collects data and
+// we need to use the newly-calculated values
+type CallbackFunc func(SlidingWindow)
+
+// slidingWindow is an object that polls for a value every `pollingInterval`
+// and then keeps those values for the duration of the `windowSize` that can
+// then be analyzed/averaged/aggregated. This object should be instantiated
+// via `NewSlidingWindow`.
+type slidingWindow struct {
+	bucketIdx       int
+	buckets         []float64
+	bucketsLock     sync.RWMutex
+	callbackFunc    CallbackFunc
+	initialized     bool
+	numBuckets      int
+	numBucketsUsed  int
+	pollingFunc     PollingFunc
+	pollingInterval time.Duration
+	stopChan        chan struct{}
+	stopped         bool
+	ticker          *time.Ticker
+	windowLock      sync.RWMutex
+	windowSize      time.Duration
+}
+
+// SlidingWindow is the public API that we expose from the slidingWindow object
+type SlidingWindow interface {
+	Start(PollingFunc, CallbackFunc) error
+	Stop()
+
+	Average() float64
+	WindowSize() time.Duration
+}
+
+// NewSlidingWindow creates a new instance of a slidingWindow
+func NewSlidingWindow(windowSize time.Duration, pollingInterval time.Duration) (SlidingWindow, error) {
+	if windowSize == 0 {
+		return nil, fmt.Errorf("SlidingWindow windowSize cannot be 0")
+	}
+
+	if pollingInterval == 0 {
+		return nil, fmt.Errorf("SlidingWindow pollingInterval cannot be 0")
+	}
+
+	if pollingInterval > windowSize {
+		return nil, fmt.Errorf("SlidingWindow windowSize must be smaller than the polling interval")
+	}
+
+	if windowSize%pollingInterval != 0 {
+		return nil, fmt.Errorf("SlidingWindow windowSize must be a multiple of polling interval")
+	}
+
+	return &slidingWindow{
+		pollingInterval: pollingInterval,
+		windowSize:      windowSize,
+	}, nil
+}
+
+// Start creates a new sliding window object, validates the parameters,
+// and starts the ticker. We use `Start` to define most of the variables
+// instead of the constructor as we are likely to need the SlidingWindow
+// instance in the polling/callback closures, leading to a chicken/egg
+// problem in usage.
```

Comment:
For `callbackFunc` the SlidingWindow is passed to the callback.  If the same were done for `pollingFunc`, would this issue be eliminated?